File: app/src/main/java/com/ruideraj/secretelephant/contacts/ContactListFragment.java
package com.ruideraj.secretelephant.contacts;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModelProviders;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.ruideraj.secretelephant.AppLog;
import com.ruideraj.secretelephant.R;

import java.util.List;

public class ContactListFragment extends Fragment implements
        ContactAdapter.ContactClickListener, View.OnClickListener {

    private static final String ARG_CONTACT_TYPE = "contactType";

    private ContactsViewModel mViewModel;
    private int mType;
    private RecyclerView mRecycler;
    private ContactAdapter mAdapter;

    public static ContactListFragment newInstance(int contactType) {
        Bundle args = new Bundle();
        args.putInt(ARG_CONTACT_TYPE, contactType);

        ContactListFragment fragment = new ContactListFragment();
        fragment.setArguments(args);

        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        AppLog.d("ContactListFragment", "onCreateView");
        mViewModel = ViewModelProviders.of(getActivity()).get(ContactsViewModel.class);

        Bundle args = getArguments();

        View root = inflater.inflate(R.layout.fragment_contact_list, container, false);

        mRecycler = root.findViewById(R.id.contacts_list_recycler);
        LinearLayoutManager manager = new LinearLayoutManager(getContext());
        mRecycler.setLayoutManager(manager);
        mAdapter = new ContactAdapter(mViewModel, this);
        mRecycler.setAdapter(mAdapter);

        mType = args.getInt(ARG_CONTACT_TYPE);

        LiveData<List<Contact>> liveData;
        if(mType == Contact.TYPE_PHONE) {
            liveData = mViewModel.phones;
        }
        else {
            liveData = mViewModel.emails;
        }

        liveData.observe(getActivity(), list -> {
            if(list != null) {
                mAdapter.setData(list);
            }
        });

        if(mType == Contact.TYPE_EMAIL) {
            if(TextUtils.isEmpty(mViewModel.emailAccount.getValue())) {
                root.findViewById(R.id.contacts_list_email_overlay).setVisibility(View.VISIBLE);
                root.findViewById(R.id.contacts_list_email_select).setOnClickListener(this);
            }
            mViewModel.emailAccount.observe(getActivity(),
                    email -> setOverlayVisible(TextUtils.isEmpty(email)));
        }

        return root;
    }

    @Override
    public void onClick(View v) {
        switch(v.getId()) {
            case R.id.contacts_list_email_select:
                mViewModel.onSelectAccount();
                break;
        }
    }

    @Override
    public void onContactClick(int position) {
        mViewModel.onContactClicked(mType, position);
        mAdapter.notifyItemChanged(position);
    }

    public void setOverlayVisible(boolean visible) {
        View root = getView();
        if(root != null) {
            View overlay = root.findViewById(R.id.contacts_list_email_overlay);
            if(visible) {
                mRecycler.setVisibility(View.GONE);
                overlay.setVisibility(View.VISIBLE);
            }
            else {
                mRecycler.setVisibility(View.VISIBLE);
                overlay.setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/java/com/ruideraj/secretelephant/contacts/ContactListFragment.kt
package com.ruideraj.secretelephant.contacts

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ruideraj.secretelephant.R
import com.ruideraj.secretelephant.ViewModelFactory
import kotlinx.coroutines.flow.collect

class ContactListFragment : Fragment() {

    companion object {
        const val ARG_CONTACT_TYPE = "contactType"

        @JvmStatic
        fun newInstance(contactType: Contact.Type): ContactListFragment {
            val args = Bundle()
            args.putSerializable(ARG_CONTACT_TYPE, contactType)

            return ContactListFragment().apply {
                arguments = args
            }
        }
    }

    private val viewModel by activityViewModels<ContactsViewModel> {
        ViewModelFactory(requireActivity())
    }
    private lateinit var type: Contact.Type
    private lateinit var recycler: RecyclerView
    private lateinit var adapter: ContactAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val args = requireArguments()
        type = args.getSerializable(ARG_CONTACT_TYPE) as Contact.Type

        val root = inflater.inflate(R.layout.fragment_contact_list, container, false)

        recycler = root.findViewById(R.id.contacts_list_recycler)
        val layoutManager = LinearLayoutManager(requireContext())
        recycler.layoutManager = layoutManager
        adapter = ContactAdapter(this, viewModel) { clickedPosition ->
            viewModel.onContactClicked(type, clickedPosition)
        }
        recycler.adapter = adapter

        viewModel.let {
            val liveData = if (type == Contact.Type.PHONE) it.phones else it.emails
            liveData.observe(requireActivity(), { contactList ->
                if (contactList != null) {
                    adapter.setData(contactList)
                }
            })

            lifecycleScope.launchWhenStarted {
                it.contactUpdate.collect { update ->
                    if (update.type == type) {
                        adapter.notifyItemChanged(update.listPosition)
                    }
                }
            }

            if (type == Contact.Type.EMAIL) {
                root.findViewById<View>(R.id.contacts_list_email_select).setOnClickListener {
                    viewModel.onSelectAccount()
                }

                if (it.emailAccount.value.isNullOrEmpty()) {
                    root.findViewById<View>(R.id.contacts_list_email_overlay).visibility = View.VISIBLE
                }

                it.emailAccount.observe(requireActivity(), { email ->
                    setOverlayVisible(email.isNullOrEmpty())
                })
            }
        }

        return root
    }

    private fun setOverlayVisible(show: Boolean) {
        view?.run {
            val overlay = findViewById<View>(R.id.contacts_list_email_overlay)

            if (show) {
                recycler.visibility = View.GONE
                overlay.visibility = View.VISIBLE
            } else {
                recycler.visibility = View.VISIBLE
                overlay.visibility = View.GONE
            }
        }
    }
}